### PR TITLE
[img2img, inpainting] fix fp16 inference

### PR DESCRIPTION
**Pull request by patil-suraj**
Fix inference for image2image and inpainting pipeline in `fp16`.

- Make sure the the init_image and mask are put fp16
- Create noise in fp16

Should be merged after #767

### CODE REVIEW

**Review by pcuenca — COMMENTED** [review]
And end-to-end test doesn't work for me with both branches applied. I think we need to cast the inputs to both `encode` and `decode`.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 1]
```diff
@@ -335,6 +298,43 @@ def __call__(
...
+        offset = self.scheduler.config.get("steps_offset", 0)
+        init_timestep = int(num_inference_steps * strength) + offset
+        init_timestep = min(init_timestep, num_inference_steps)
+
+        timesteps = self.scheduler.timesteps[-init_timestep]
+        timesteps = torch.tensor([timesteps] * batch_size * num_images_per_prompt, device=self.device)
+
+        # add noise to latents using the timesteps
+        noise = torch.randn(init_latents.shape, generator=generator, device=self.device, dtype=latents_dtype)
+        init_latents = self.scheduler.add_noise(init_latents, noise, timesteps)
```
Comment (NouamaneTazi) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:336]:
```suggestion
        init_latents = self.scheduler.add_noise(init_latents, noise, timesteps).to(dtype=latents_dtype)
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 2]
```diff
@@ -297,6 +277,28 @@ def __call__(
...
+        offset = self.scheduler.config.get("steps_offset", 0)
+        init_timestep = int(num_inference_steps * strength) + offset
+        init_timestep = min(init_timestep, num_inference_steps)
+
+        timesteps = self.scheduler.timesteps[-init_timestep]
+        timesteps = torch.tensor([timesteps] * batch_size * num_images_per_prompt, device=self.device)
+
+        # add noise to latents using the timesteps
+        noise = torch.randn(init_latents.shape, generator=generator, device=self.device, dtype=latents_dtype)
+        init_latents = self.scheduler.add_noise(init_latents, noise, timesteps)
```
Comment (NouamaneTazi) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:300]:
```suggestion
        init_latents = self.scheduler.add_noise(init_latents, noise, timesteps).to(dtype=latents_dtype)
```

**Review by NouamaneTazi — COMMENTED** [review]
The changes I suggested seem to be already in https://github.com/huggingface/diffusers/pull/767/. But I added them to make sure they stay nonetheless

**Comment by patrickvonplaten** [comment]
Code looks good to me! If code works for @pcuenca then good to merge for me :-)

**Review by pcuenca — APPROVED** [review]
LGTM!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py (hunk 1 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:336]: This is handled in #767 , the output of `add_noise` will have same `dtype` as it's `input`, so no need to  cast here again.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py (hunk 2 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:300]: same comment as above

**Comment by pcuenca** [comment]
Tests look good too. Thanks a lot!
